### PR TITLE
simple transaction

### PR DESCRIPTION
**Pull request by Nirmalperiyasamy**
Need to add date and time  of transaction further more.

### CODE REVIEW

File: src/main/java/com/nirmal/banking/common/Const.java [hunk 1]
```diff
@@ -20,6 +20,12 @@ public class Const {
 
     public static final String VALIDATE_KYC = "/decision-status/{uid}/{decision}";
 
-    public static final String TOKEN = "/token";
+    public static final String LOGIN = "/login";
+
+    public static final String DEBIT = "/debit";
+
+    public static final String AMOUNT_BALANCE  ="/amount-balance";
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/common/Const.java]: replace with balance

File: src/main/java/com/nirmal/banking/controller/LoginController.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    private final AuthenticationManager authenticationManager;
+
+    private final JwtUtil jwtUtil;
+
+    @PostMapping(LOGIN)
+    private ResponseEntity<?> token(@RequestBody UserDetailsDto userDetailsDto, HttpServletResponse response) {
+        String uid = userService.findByName(userDetailsDto.getUsername());
+        try {
+            authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(uid, userDetailsDto.getPassword()));
+        } catch (Exception e) {
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/controller/LoginController.java:37]: wrap whole function with try catch

File: src/main/java/com/nirmal/banking/service/UserService.java [hunk 3]
```diff
@@ -52,26 +57,24 @@ public UserDetails loadUserByUsername(String uid) throws UsernameNotFoundExcepti
...
+        if (userExist(userDetailsDto.getUsername())) {
+            throw new CustomException(ErrorMessages.USERNAME_EXIST);
+        }
         CustomUserDetails customUserDetails = new CustomUserDetails();
         BeanUtils.copyProperties(userDetailsDto, customUserDetails);
         customUserDetails.setUserRole(roleService.getRole(Role.USER));
         customUserDetails.setPassword(passwordEncoder.encode(userDetailsDto.getPassword()));
         customUserDetails.setUid(UUID.randomUUID().toString());
-        customUserDetails.setKycStatus(KycStatus.PENDING);
+        customUserDetails.setKycStatus(KycStatus.APPROVED);
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/service/UserService.java]: Why is it by default APPROVED? Is there any reasons for it

File: src/main/java/com/nirmal/banking/common/Const.java (hunk 1 above)
Comment (Nirmalperiyasamy) [reply src/main/java/com/nirmal/banking/common/Const.java]: ok

File: src/main/java/com/nirmal/banking/service/UserService.java [hunk 4]
```diff
@@ -84,6 +87,57 @@ public String uploadImage(MultipartFile[] file, HttpServletRequest request) thro
         return SuccessMessages.UPLOADED;
     }
 
+    public String depositAmount(HttpServletRequest request, Integer depositAmount) {
+        String uid = extractUid(request);
+        CustomUserDetails customUserDetails = userDetailsRepo.findByuid(uid);
+        if (!customUserDetails.getKycStatus().equals(KycStatus.APPROVED)) {
+            throw new CustomException(ErrorMessages.KYC_NOT_APPROVED);
+        }
+        if (!transactionRepo.existsByUid(uid)) {
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/service/UserService.java]: add comment

File: src/main/java/com/nirmal/banking/service/UserService.java [hunk 5]
```diff
@@ -84,6 +87,57 @@ public String uploadImage(MultipartFile[] file, HttpServletRequest request) thro
...
+        }
+        TransactionDetails transactionDetails = new TransactionDetails();
+        TransactionDetailsDto transactionDetailsDto = TransactionDetailsDto.builder()
+                .uid(uid)
+                .amount(depositAmount)
+                .build();
+        BeanUtils.copyProperties(transactionDetailsDto, transactionDetails);
+        transactionRepo.save(transactionDetails);
+        transactionDetails.setTotalAmount(transactionRepo.getTotalAmountByUid(uid));
+        transactionRepo.save(transactionDetails);
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/service/UserService.java:105]: could use some refactors

File: src/main/java/com/nirmal/banking/controller/UserController.java [hunk 6]
```diff
@@ -30,28 +21,29 @@ public class UserController {
...
-        } catch (Exception e) {
-            throw new CustomException(ErrorMessages.INVALID);
-        }
-        return ResponseEntity.ok(jwtUtil.generateToken(uid));
+    @PostMapping(DEPOSIT)
+    private ResponseEntity<?> deposit(@RequestBody TransactionDetailsDto transactionDetailsDto, HttpServletRequest request){
+        return ResponseEntity.ok(userService.depositAmount(request,transactionDetailsDto.getAmount()));
+    }
+
+    @PostMapping(DEBIT)
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/controller/UserController.java]: Can be changed to withdraw

File: src/main/java/com/nirmal/banking/service/UserService.java [hunk 7]
```diff
@@ -84,8 +85,70 @@ public String uploadImage(MultipartFile[] file, HttpServletRequest request) thro
...
+        CustomUserDetails customUserDetails = userDetailsRepo.findByuid(uid);
 
+        if (!customUserDetails.getKycStatus().equals(KycStatus.APPROVED))
+            throw new CustomException(ErrorMessages.KYC_NOT_APPROVED);
+
+        if (!transactionRepo.existsByUid(uid))
+            if (depositAmount < 1000) throw new CustomException(ErrorMessages.MINIMUM_DEPOSIT_AMOUNT);
+
+        TransactionDetails transactionDetails = new TransactionDetails();
+        TransactionDetailsDto transactionDetailsDto = TransactionDetailsDto.builder()
```
Comment (gokul656) [inline src/main/java/com/nirmal/banking/service/UserService.java]: why do we need DTO here?

**Review by gokul656 — CHANGES_REQUESTED** [review]

File: src/main/java/com/nirmal/banking/service/UserService.java (hunk 7 above)
Comment (Nirmalperiyasamy) [reply src/main/java/com/nirmal/banking/service/UserService.java]: we are inserting data into  table , through dto we done the process.

**Review by gokul656 — APPROVED** [review]

**Pull request merged**